File: 30-Day LeetCoding Challenge/Python/maximalSquare.py
class Solution(object):
    def maximalSquare(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        ans = 0
        if not matrix:
            return ans
        
        rows = len(matrix)
        cols = len(matrix[0])
        dp = [[0 for i in range(cols)] for j in range(rows)]
        for row in range(rows):
            for col in range(cols):
                if matrix[row][col] == '1':
                    dp[row][col] = 1
                    if row > 0 and col > 0:
                        dp[row][col] += min(dp[row][col-1], dp[row-1][col], dp[row-1][col-1])
                ans = max(ans, dp[row][col])
        return ans*ans

File: 30-Day LeetCoding Challenge/C++/maximalSquare.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        if(matrix.size() == 0 || matrix[0].size() == 0) {
            return 0;
        }
        int ans = 0;
        int rows = matrix.size();
        int cols = matrix[0].size();
        vector<vector<int>> dp(rows, vector<int>(cols));
        for(int row = 0; row < rows; row++) {
            for(int col = 0; col < cols; col++) {
                if(matrix[row][col] == '1') {
                    dp[row][col] = 1;
                    if(row > 0 && col > 0) {
                    dp[row][col] += min({dp[row][col-1], dp[row-1][col], dp[row-1][col-1]});
                    }
                    
                }
                
                ans = max(ans, dp[row][col]);
            }
        }
        return ans * ans ;
    }
};